### PR TITLE
add ability to disable streaming http request processing

### PR DESCRIPTION
**Pull request by graemerocher**
Adds a new `micronaut.server.netty.server-type` setting that can be set to `FULL_CONTENT` to disable streaming requests. This will make it easier to implement certain use cases like retrieving the body from a filter and also improves performance.

This PR also optimises route execution further to improvement the overall performance of the framework.

### CODE REVIEW

**Comment by yawkat** [comment]
the tests pass now, i had to make a bunch of changes unfortunately to make all the features work. eg i reverted NettyHttpRequest to still use the old content/data approach.

i'm going to do some profiling to see if there are new simple optimizations

File: core/src/main/java/io/micronaut/core/bind/OneShotBindingResult.java
Comment (dstepanov) [inline core/src/main/java/io/micronaut/core/bind/OneShotBindingResult.java]: This shouldn't be needed after changes in my PR

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateMultiObjectBody.java
Comment (dstepanov) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateMultiObjectBody.java]: You might want to use CollectionUtils.newLinkedHashMap to create a map of size dataList.size

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/multipart/MultipartBodyArgumentBinder.java
Comment (dstepanov) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/multipart/MultipartBodyArgumentBinder.java]: My PR doesn't use a custom subscriber as it isn't needed just a simple filter/map

File: http/src/main/java/io/micronaut/http/bind/binders/DefaultBodyAnnotationBinder.java
Comment (dstepanov) [inline http/src/main/java/io/micronaut/http/bind/binders/DefaultBodyAnnotationBinder.java]: How will it work when HttpRequest is received in the controller and the body is read?

File: core/src/main/java/io/micronaut/core/bind/OneShotBindingResult.java
Comment (dstepanov) [inline core/src/main/java/io/micronaut/core/bind/OneShotBindingResult.java]: Please add all since, author to the new files

**Review by dstepanov — CHANGES_REQUESTED** [review]

**Comment by yawkat** [comment]
@dstepanov re changes from #9017. i will merge #9006 and #9017 once #9006 is merged. if i merge before then, i would have to redo the merge since the changes in #9006 will be squashed.

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java
Comment (dstepanov) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java:177]: Why is this looping? I see it as one or nothing

Comment (dstepanov) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java:177]: I guess because of the missing subscription

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java
Comment (dstepanov) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java:83]: I would implement it as a blocking queue to reduce the locking code

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java
Comment (dstepanov) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java:177]: I would implement this as `boolean consumeBlocking(Consumer c)` to handle the release in the same class

**Review by dstepanov — CHANGES_REQUESTED** [review]

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java
Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java:177]: yea it's because of the subscription, and also because await() can have a spurious wakeup.

can't do release in the same method, because PublisherAsStream stores the bytebuf for a time in case the read output buffer is too small

Comment (dstepanov) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java:177]: Ok

File: http/src/main/java/io/micronaut/http/bind/binders/DefaultBodyAnnotationBinder.java
Comment (yawkat) [reply http/src/main/java/io/micronaut/http/bind/binders/DefaultBodyAnnotationBinder.java]: same as before. i think this comment relates to the changes ive just reverted.

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java
Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java:83]: blockingqueue would not work i think. there would be a race condition with setting the subscription field. checking the subscription field and blocking on the result must happen in one atomic step, i dont think it's possible with blockingqueue.

File: core/src/main/java/io/micronaut/core/bind/OneShotBindingResult.java
Comment (yawkat) [reply core/src/main/java/io/micronaut/core/bind/OneShotBindingResult.java]: oh this became unused anyway apparently

**Comment by yawkat** [comment]
idk what's wrong with sonar coverage, it shows missing coverage for some code that is clearly covered. e.g. the code for writing a non-streaming response is shown as uncovered. but i wrote that code specifically to fix tests...

**Review by dstepanov — APPROVED** [review]

File: http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java
Comment (JLLeitschuh) [inline http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java:286]: Why is header validation disabled here?

Related: https://github.com/micronaut-projects/micronaut-core/security/advisories/GHSA-694p-xrhg-x3wm

File: http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java
Comment (JLLeitschuh) [inline http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java:310]: Similar question to the above, why is header validation disabled here?

File: http-netty/src/main/java/io/micronaut/http/netty/NettyMutableHttpResponse.java
Comment (JLLeitschuh) [inline http-netty/src/main/java/io/micronaut/http/netty/NettyMutableHttpResponse.java:311]: Why do this validation here? Why not put this logic in `NettyHttpHeaders` so that there exists no API on `NettyHttpHeaders` that allows for an unsafe use?

File: http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java
Comment (graemerocher) [reply http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java:286]: The headers were already validated in `headers.getNettyHeaders()` so we know it is safe

File: http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java
Comment (graemerocher) [reply http-client/src/main/java/io/micronaut/http/client/netty/NettyClientHttpRequest.java:310]: same answer as above

File: http-netty/src/main/java/io/micronaut/http/netty/NettyMutableHttpResponse.java
Comment (graemerocher) [reply http-netty/src/main/java/io/micronaut/http/netty/NettyMutableHttpResponse.java:311]: these are internal non-public APIs inaccessible to users, this PR was partly to optimise performance, so excessive validation when it is not needed is not great for performance.

**Pull request merged**
